src/ApparentHorizons/SpherePackIterator.hpp

```diff
@@ -0,0 +1,106 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+///
+/// \file
+/// Defines class SpherePackIterator.
+
+//
+// iterate through an array with the layout of the Ylm-spectral
+// coefficients of SpherePack.
+//================================================================
+#pragma once
+
+#include <ostream>
+#include <vector>
+
+/// Iterates through the coefficients of a SpherePack spectral
+/// transform.  The internal SpherePack ordering is not intuitive, so
+/// SpherePackIterator exists for the purpose of iterating over the
+/// array and determining which specific values of l and m correspond
+/// to each iteration.
+///
+/// \detail
+/// SpherePackIterator iterates over each coefficient ONCE, in some
+/// arbitrary order. There are member functions to determine the values
+/// of l and m at each iteration.  Also, operator() returns the position
+/// in the array at each iteration; this position can be used to index
+/// vectors that correspond to spectral coefficients stored
+/// in SpherePack format.
+/// \example
+/// \snippet Test_SpherePackIterator.cpp spherepack_iterator_example
+class SpherePackIterator {
+ public:
+  enum class CoefficientArray { a, b };
+  SpherePackIterator(size_t l_max_input, size_t m_max_input, size_t stride = 1);
+
+  size_t l_max() const { return l_max_; }
+  size_t m_max() const { return m_max_; }
+  size_t n_th() const { return n_th_; }
+  size_t n_ph() const { return n_ph_; }
+
+  /// Size of the combined a/b array
+  size_t array_size() const { return (l_max_ + 1) * (m_max_ + 1) * 2; }
+
+  SpherePackIterator& operator++() {
+    if (index_ == index_array_.size()) {
```

Comment:
why is this not an ASSERT?

If it cannot be an ASSERT, if(index_ < ...) { ++index_;} return *this;